test/Analyzers.Tests/Orchestration/DateTimeOrchestrationAnalyzerTests.cs

```diff
@@ -0,0 +1,331 @@
+
+namespace Microsoft.DurableTask.Analyzers.Tests.Orchestration;
+
+public class DateTimeOrchestrationAnalyzerTests
+{
+    [Fact]
+    public async Task EmptyCodeWithNoSymbolsAvailableHasNoDiag()
+    {
+        string code = @"";
+
+        // checks that empty code with no assembly references of Durable Functions has no diagnostics.
+        // this guarantees that if someone adds our analyzer to a project that doesn't use Durable Functions,
+        // the analyzer won't crash/they won't get any diagnostics
+        await VerifyCS.VerifyAnalyzerAsync(code);
+    }
+
+    [Fact]
+    public async Task EmptyCodeWithSymbolsAvailableHasNoDiag()
+    {
+        string code = @"";
+
+        // checks that empty code with access to assembly references of Durable Functions has no diagnostics
+        await VerifyCS.VerifyDurableTaskAnalyzerAsync(code);
+    }
+
+    [Fact]
+    public async Task NonOrchestrationHasNoDiag()
+    {
+        string code = WrapDurableFunctionOrchestration(@"
+void Method(){
+    Console.WriteLine(DateTime.Now);
+}
+");
+
+        await VerifyCS.VerifyDurableTaskAnalyzerAsync(code);
+    }
+
+    [Theory]
+    [InlineData("DateTime.Now")]
+    [InlineData("DateTime.UtcNow")]
+    [InlineData("DateTime.Today")]
+    public async Task DurableFunctionOrchestrationUsingDateTimeNonDeterministicPropertiesHasDiag(string expression)
+    {
+        string code = WrapDurableFunctionOrchestration($@"
+[Function(""Run"")]
+DateTime Run([OrchestrationTrigger] TaskOrchestrationContext context)
+{{
+    return {{|#0:{expression}|}};
+}}
+");
+
+        DiagnosticResult expected = BuildDiagnostic().WithLocation(0).WithArguments("Run", $"System.{expression}", "Run");
+
+        await VerifyCS.VerifyDurableTaskAnalyzerAsync(code, expected);
+    }
+
+    [Fact]
+    public async Task DurableFunctionOrchestrationInvokingChainedMethodsHasDiag()
+    {
+        string code = WrapDurableFunctionOrchestration(@"
+[Function(""Run1"")]
+long Run1([OrchestrationTrigger] TaskOrchestrationContext context)  => Level1();
+
+[Function(""Run2"")]
+long Run2([OrchestrationTrigger] TaskOrchestrationContext context)  => Level1();
+
+long Level1() => Level2();
+
+long Level2() => Level3();
+
+long Level3() => {|#0:DateTime.Now|}.Ticks;
+");
+
+        DiagnosticResult expected = BuildDiagnostic().WithLocation(0).WithArguments("Level3", "System.DateTime.Now", "Run1, Run2");
+
+        await VerifyCS.VerifyDurableTaskAnalyzerAsync(code, expected);
+    }
+
+    [Fact]
+    public async Task DurableFunctionOrchestrationInvokingRecursiveMethodsHasSingleDiag()
+    {
+        string code = WrapDurableFunctionOrchestration(@"
+[Function(""Run"")]
+long Run([OrchestrationTrigger] TaskOrchestrationContext context) => RecursiveMethod(0);
+
+long RecursiveMethod(int i){
+    if (i == 10) return 1;
+    DateTime date = {|#0:DateTime.Now|};
```

Comment:
Interesting we will examine called methods. How far do we look? How complex of a call chain can this follow? What happens if it is an interface or abstraction method being called?

Reply:
> What is the expected behavior of the below code:

The `DateTime.Now` property reference operation would have the following diagnostic:
`"The method 'HelperMethod' uses 'System.Date.Now' that may cause non-deterministic behavior when invoked from orchestration 'MyOrchestrator'"`

Within the diagnostic message, I tried to explicitly state that the helper method is not the problem per say, but instead, **its invocation from an orchestration**. That's why for all types of orchestration I try to locate the function name and add it to the diagnostic message. As soon as you remove the call from the orchestration, the diagnostic will disappear.

> How far do we look? How complex of a call chain can this follow? What happens if it is an interface or abstraction method being called?

I am glad you brought this case up, I haven't thought about it before😁. I just checked and we can follow the call chain of every method that has an implementation that isn't hidden behind an interface, abstract class or a library which we don't have access to the actual syntax nodes. For instance, let's say you have:

```cs
interface IHelper {
    DateTime Method();
}

class Helper : IHelper {
    public DateTime Method() => DateTime.Now;
}

class Orchestrator
{
    // this triggers the diagnostic
    [Function(""Run"")]
    void Run([OrchestrationTrigger] TaskOrchestrationContext context)
    {
        var helper = new Helper();
        _ = helper.Method();
    }

    // this does not trigger the diagnostic
    [Function(""Run"")]
    void Run([OrchestrationTrigger] TaskOrchestrationContext context)
    {
        IHelper helper = new Helper();
        _ = helper.Method();
    }
}
```

> I think it may be valuable for us to formally detail how our analyzers will follow method call chains that originate in an orchestration. How will this behave with static, instance, interfaces, abstract, etc methods? And as we create this spec we should consider if this could impact non-orchestration scenarios.

Reply:
>  By formally detail, do you mean writing about that behavior in our future analyzer docs?

Write about it in docs, but also write it for us to follow ourselves as we add new analyzers. The method probing behavior should be consistent between all our analyzers.

> I just checked and we can follow the call chain of every method...

Is there a perf concern here? What happens when there is a recursive method?

Reply:
> Write about it in docs, but also write it for us to follow ourselves as we add new analyzers.

Great, I will make sure to document it before we release it: https://github.com/microsoft/durabletask-dotnet/issues/286

> Is there a perf concern here? What happens when there is a recursive method?

Not actually, there is a guard against it: https://github.com/microsoft/durabletask-dotnet/blob/4c97c965db857128f1e737f6888d7513c60d4e82/src/Analyzers/Orchestration/OrchestrationAnalyzer.cs#L198-L203

Reply:
Performance wise, I am going to benchmark it soon 👍